go/vt/vttablet/tabletmanager/vreplication/framework_test.go

```diff
@@ -118,17 +122,19 @@ func cleanup() {
 	playerEngine.Close()
 	streamerEngine.Close()
 	env.Close()
+	envMu.Unlock()
 }
 
 func setup() (func(), int) {
-	globalDBQueries = make(chan string, 1000)
-	resetBinlogClient()
 	var err error
 	env, err = testenv.Init()
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "%v", err)
 		return nil, 1
 	}
+	envMu.Lock()
```

Comment:
I am not clear how the mutex is helping. What is it serializing now that wasn't earlier? The code, before I commented out the noblob set is shown below. We were calling setup and cancel sequentially and waiting for each to complete. 
```
                cancel()
		runNoBlobTest = true
		if err := utils.SetBinlogRowImageMode("noblob", tempDir); err != nil {
			panic(err)
		}
		defer utils.SetBinlogRowImageMode("", tempDir)
		cancel, ret = setup()
		if ret > 0 {
			return ret
		}
		defer cancel()
		ret = m.Run()
```

Reply:
Remember that the failures are ONLY seen with `go test -race` via the `unit race` workflow AFAIK. The race detector is described here: https://go.dev/doc/articles/race_detector

It modifies to the code to intentionally try and create timing related issues and it reports when two coroutines access the same variable concurrently and at least one is a write.

The race detector uses CGO and modifies the program to try and suss out these potential data races. The only thing I can proffer here is that it is allowing things not explicitly enforced in order, by synchronization primitives, to be executed out of order (from the source code order), as can happen on weak memory model systems (since the go runtime is not tied to any CPU arch). And somehow we then started the next test (m.Run()) while the first one was still running. That, or it may be that the use of the new mutex here and the cache coherency round from that is enough to allow the post copy phase go routine to finish.